src/software/simulation/physics_simulator_robot.h

```diff
@@ -0,0 +1,311 @@
+    void brakeMotorBackLeft() override;
+    void brakeMotorBackRight() override;
+    void brakeMotorFrontLeft() override;
+    void brakeMotorFrontRight() override;
+
+    /**
+     * Sets the current primitive this robot is running to a new one
+     *
+     * @param firmware_world The world to run the primitive in
+     * @param primitive_msg The primitive to start
+     */
+    void startNewPrimitive(std::shared_ptr<FirmwareWorld_t> firmware_world,
+                           const TbotsProto_Primitive& primitive_msg) override;
+
+    /**
+     * Runs the current primitive
+     *
+     * @param world The world to run the primitive in
+     */
+    void runCurrentPrimitive(std::shared_ptr<FirmwareWorld_t> firmware_world) override;
+
+   private:
+    /**
+     * A function that is called during every physics step for as long as the ball
+     * is touching this robot's dribbler
+     *
+     * @param physics_robot The robot involved in the contact
+     * @param physics_ball The ball invovled in the contact
+     */
+    void onDribblerBallContact(PhysicsRobot* physics_robot, PhysicsBall* physics_ball);
+
+    /**
+     * A function that is called during once when the ball starts touching
+     * this robot's dribbler.
+     *
+     * @param physics_robot The robot involved in the contact
+     * @param physics_ball The ball invovled in the contact
+     */
+    void onDribblerBallStartContact(PhysicsRobot* physics_robot,
+                                    PhysicsBall* physics_ball);
+
+    /**
+     * A function that is called during once when the ball stops touching
+     * this robot's dribbler.
+     *
+     * @param physics_robot The robot involved in the contact
+     * @param physics_ball The ball invovled in the contact
+     */
+    void onDribblerBallEndContact(PhysicsRobot* physics_robot, PhysicsBall* physics_ball);
+
+    /**
+     * Helper functions that check if the current pointer to the physics_robot is valid
+     * before calling the given function. If the physics_robot is invalid, a warning is
+     * logged and a default value is returned.
+     *
+     * @param func The function to perform on the physics robot
+     */
+    void checkValidAndExecuteVoid(
+        std::function<void(std::shared_ptr<PhysicsRobot>)> func);
+    float checkValidAndReturnFloat(
+        std::function<float(std::shared_ptr<PhysicsRobot>)> func);
+    unsigned int checkValidAndReturnUint(
+        std::function<unsigned int(std::shared_ptr<PhysicsRobot>)> func);
+
+    /**
+     * Applies force to the physics ball to simulate it being dribbled by the
+     * physics robot.
+     *
+     * @param physics_robot The robot that should dribble the ball
+     * @param physics_ball The ball to be dribbled
+     */
+    void applyDribblerForce(PhysicsRobot* physics_robot, PhysicsBall* physics_ball);
+
+    std::weak_ptr<PhysicsRobot> physics_robot;
+    std::optional<float> autokick_speed_m_per_s;
+    std::optional<float> autochip_distance_m;
```

Comment:
would this be worth putting in the base class?

Reply:
Moved to base class